File: nuwa-script/implementations/python/nuwa/ast.py
import dataclasses
from typing import Any, Dict, List, Optional, Union

# Base node
@dataclasses.dataclass
class Node:
    """Base class for all AST nodes."""
    pass

# Expressions
@dataclasses.dataclass
class Expression(Node):
    """Base class for all expression nodes."""
    pass

@dataclasses.dataclass
class Literal(Expression):
    """Represents literal values like numbers, strings, booleans."""
    value: Union[int, float, str, bool, None]

@dataclasses.dataclass
class Variable(Expression):
    """Represents a variable identifier."""
    name: str

@dataclasses.dataclass
class BinaryOp(Expression):
    """Represents a binary operation (e.g., price > 70000, a AND b)."""
    operator: str  # e.g., "==", "!=", ">", "<", ">=", "<=", "AND", "OR"
    left: Expression
    right: Expression

@dataclasses.dataclass
class UnaryOp(Expression):
    """Represents a unary operation (e.g., NOT condition)."""
    operator: str # e.g., "NOT"
    operand: Expression

@dataclasses.dataclass
class FunctionCall(Expression):
    """Represents a built-in function call (e.g., NOW())."""
    function_name: str
    # Currently no arguments for NOW(), but might need in future
    # arguments: List[Expression] = dataclasses.field(default_factory=list)

@dataclasses.dataclass
class CallExpression(Expression):
    """Represents a tool call used as an expression (e.g., in LET)."""
    tool_name: str
    arguments: Dict[str, Expression]

@dataclasses.dataclass
class CalcExpression(Expression):
    """Represents a CALC block."""
    formula: str
    variables: Dict[str, Expression]


# Statements
@dataclasses.dataclass
class Statement(Node):
    """Base class for all statement nodes."""
    pass

@dataclasses.dataclass
class LetStatement(Statement):
    """Represents a LET statement."""
    variable_name: str
    value: Expression # Can be Literal, Variable, CallExpression, CalcExpression, etc.

@dataclasses.dataclass
class CallStatement(Statement):
    """Represents a CALL statement."""
    tool_name: str
    arguments: Dict[str, Expression]

@dataclasses.dataclass
class IfStatement(Statement):
    """Represents an IF statement."""
    condition: Expression
    then_block: List[Statement]
    else_block: Optional[List[Statement]] = None # ELSE block is optional

@dataclasses.dataclass
class ForStatement(Statement):
    """Represents a FOR statement."""
    iterator_variable: str
    iterable: Expression # The list/collection to iterate over
    loop_block: List[Statement]

@dataclasses.dataclass
class Script(Node):
    """Represents the entire script (a list of statements)."""
    statements: List[Statement]

File: nuwa-script/implementations/typescript/src/ast.ts
/**
 * Represents the different types of runtime values NuwaScript deals with.
 * This is defined more concretely in values.ts, but needed here for Literal types.
 */
export type NuwaValue =
  | string
  | number
  | boolean
  | null
  | NuwaValue[]
  | { [key: string]: NuwaValue };

// --- Base Nodes ---

export interface BaseNode {
  // Optional: Add location info (start/end line/column) from parser later
  // location?: { start: Position, end: Position };
}

// --- Expressions ---

export type Expression =
  | LiteralExpr
  | VariableExpr
  | BinaryOpExpr
  | UnaryOpExpr
  | FunctionCallExpr
  | ToolCallExpr
  | ArrayIndexExpression
  | MemberAccessExpr;
  // | CalcExpr; // Removed CALC

export interface LiteralExpr extends BaseNode {
  kind: 'LiteralExpr';
  value: NuwaValue;
}

export interface VariableExpr extends BaseNode {
  kind: 'VariableExpr';
  name: string; // Can be simple "x" or dotted "obj.prop"
}

export type BinaryOperator =
  | '==' | '!=' | '>' | '<' | '>=' | '<='
  | 'AND' | 'OR'
  | '+' | '-' | '*' | '/'; // Add arithmetic

export interface BinaryOpExpr extends BaseNode {
  kind: 'BinaryOpExpr';
  operator: BinaryOperator;
  left: Expression;
  right: Expression;
}

export type UnaryOperator = 'NOT' | '+' | '-'; // Add PLUS and MINUS

export interface UnaryOpExpr extends BaseNode {
  kind: 'UnaryOpExpr';
  operator: UnaryOperator;
  operand: Expression;
}

export type BuiltinFunctionName = 'NOW';

export interface FunctionCallExpr extends BaseNode {
  kind: 'FunctionCallExpr';
  functionName: BuiltinFunctionName;
  // arguments: Expression[]; // NOW() has no args, PRINT is a statement
}

// Represents CALL tool { ... } used as an expression
export interface ToolCallExpr extends BaseNode {
  kind: 'ToolCallExpr';
  toolName: string;
  arguments: Record<string, Expression>; // Map arg name to its expression
}

// Add the new ArrayIndexExpression interface
export interface ArrayIndexExpression extends BaseNode {
  kind: 'ArrayIndexExpression';
  object: Expression; // The expression being indexed (e.g., the variable name)
  index: Expression; // The expression within the brackets
}

// Add the new MemberAccessExpr interface
export interface MemberAccessExpr extends BaseNode {
  kind: 'MemberAccessExpr';
  object: Expression; // The expression whose member is accessed (e.g., variable or result of another expr)
  property: string;   // The name of the property being accessed
}

// Represents CALC { formula: "...", vars: {...} } - REMOVED
/*
export interface CalcExpr extends BaseNode {
  kind: 'CalcExpr';
  formula: string;
  variables: Record<string, Expression>; // Map internal var name to its expression
}
*/

// --- Statements ---

export type Statement =
  | LetStatement
  | CallStatement
  | IfStatement
  | ForStatement
  | PrintStatement; // Include PRINT here as a statement

export interface LetStatement extends BaseNode {
  kind: 'LetStatement';
  variableName: string;
  value: Expression;
}

export interface CallStatement extends BaseNode {
  kind: 'CallStatement';
  toolName: string;
  arguments: Record<string, Expression>;
}

export interface IfStatement extends BaseNode {
  kind: 'IfStatement';
  condition: Expression;
  thenBlock: Statement[]; // Body is a list of statements
  elseBlock?: Statement[]; // Optional else block
}

export interface ForStatement extends BaseNode {
  kind: 'ForStatement';
  iteratorVariable: string;
  iterable: Expression; // Expression evaluating to the list
  loopBlock: Statement[]; // Body is a list of statements
}

export interface PrintStatement extends BaseNode {
  kind: 'PrintStatement';
  value: Expression; // The expression to evaluate and print
}

// --- Top Level ---

export interface Script extends BaseNode {
  kind: 'Script';
  statements: Statement[];
}

// --- Helper Functions (Type Guards) ---
// These help in the interpreter to narrow down the type

export function isLiteralExpr(node: BaseNode): node is LiteralExpr {
    return (node as any)?.kind === 'LiteralExpr';
}

export function isVariableExpr(node: BaseNode): node is VariableExpr {
    return (node as any)?.kind === 'VariableExpr';
}

export function isBinaryOpExpr(node: BaseNode): node is BinaryOpExpr {
    return (node as any)?.kind === 'BinaryOpExpr';
}

// Add the type guard for ArrayIndexExpression
export function isArrayIndexExpression(node: BaseNode): node is ArrayIndexExpression {
    return (node as any)?.kind === 'ArrayIndexExpression';
}

// Add the type guard for MemberAccessExpr
export function isMemberAccessExpression(node: BaseNode): node is MemberAccessExpr {
    return (node as any)?.kind === 'MemberAccessExpr';
}

export function isLetStatement(node: BaseNode): node is LetStatement {
     return (node as any)?.kind === 'LetStatement';
}

export function isCallStatement(node: BaseNode): node is CallStatement {
     return (node as any)?.kind === 'CallStatement';
}

export function isIfStatement(node: BaseNode): node is IfStatement {
     return (node as any)?.kind === 'IfStatement';
}

export function isForStatement(node: BaseNode): node is ForStatement {
     return (node as any)?.kind === 'ForStatement';
}

export function isPrintStatement(node: BaseNode): node is PrintStatement {
     return (node as any)?.kind === 'PrintStatement';
}
